File: Server_Root/Server/RoomServer/Entity/Entities/Projectile/Projectile.cpp
#include "stdafx.h"
#include "Projectile.h"
#include "../../../MasterData/MasterDataManager.h"
#include "../../../MasterData/Projectile.h"
#include "../../../MasterData/Behavior.h"
#include "../../../Factory.h"
#include "../../../Util.h"
#include "../../../GameEvent/GameEvents/BehaviorEnd.h"
#include "../../../Game/BaeGameRoom.h"
#include "../Character/Character.h"

Projectile::Projectile(BaeGameRoom* pGameRoom, int nProjectorID, int nID, int nMasterDataID) : IEntity(pGameRoom, nID, nMasterDataID)
{
	m_nProjectorID = nProjectorID;
	m_Projector = (Character*)m_pGameRoom->GetEntity(nProjectorID);
}

Projectile::~Projectile()
{
}

void Projectile::Initialize()
{
	MasterData::Projectile* pMasterProjectile = NULL;
	MasterDataManager::Instance()->GetData<MasterData::Projectile>(m_nMasterDataID, pMasterProjectile);

	m_fSize = pMasterProjectile->m_fSize;
	m_fHeight = pMasterProjectile->m_fHeight;
	m_fDefault_Y = pMasterProjectile->m_fDefault_Y;
	m_nDefaultBehaviorID = pMasterProjectile->m_nDefaultBehaviorID;

	for (vector<int>::iterator it = pMasterProjectile->m_vecBehaviorID.begin(); it != pMasterProjectile->m_vecBehaviorID.end(); ++it)
	{
		IBehavior* pBehavior = Factory::Instance()->CreateBehavior(m_pGameRoom, this, *it);
		if (pBehavior != NULL)
		{
			pBehavior->Initialize();
			m_listBehavior.push_back(pBehavior);
		}
	}
	Util::Parse(pMasterProjectile->m_strLifeInfo, ':', &m_vecLifeInfo);
}

float Projectile::GetMoveSpeed()
{
	return 3;
}

FBS::Data::EntityType Projectile::GetEntityType()
{
	return FBS::Data::EntityType::EntityType_Projectile;
}

void Projectile::NotifyGameEvent(IGameEvent* pGameEvent)
{
	if (m_vecLifeInfo[0] == "Arrival" && pGameEvent->GetType() == FBS::GameEventType::GameEventType_BehaviorEnd)
	{
		GameEvent::BehaviorEnd* pBehaviorEnd = (GameEvent::BehaviorEnd*)pGameEvent;

		MasterData::Behavior* pMasterBehavior = NULL;
		MasterDataManager::Instance()->GetData<MasterData::Behavior>(pBehaviorEnd->m_nBehaviorID, pMasterBehavior);

		if (pMasterBehavior->m_strClassName == "Move")
		{
			m_pGameRoom->DestroyEntity(m_nID);
		}
	}
}

bool Projectile::IsTerrainPassable()
{
	return false;
}

int Projectile::GetMoveCollisionTypes()
{
	int nTypes = CollisionObject::Type::CollisionObjectType_None;

	if (m_Projector->GetRole() == Character::Role::Challenger)
	{
		nTypes = CollisionObject::Type::CollisionObjectType_Character;
	}
	else if (m_Projector->GetRole() == Character::Role::Disturber)
	{
		nTypes = CollisionObject::Type::CollisionObjectType_Character_Challenger;
	}

	return nTypes;
}

int Projectile::GetProjectorID()
{
	return m_nProjectorID;
}

Character* Projectile::GetProjector()
{
	return m_Projector;
}

File: Client_Root/Client/Assets/Scripts/Entity/Entities/Projectile/Projectile.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class Projectile : IEntity
{
    public override float GetSpeed()
    {
        return 0;
    }

    public override float GetMaximumSpeed()
    {
        return 0;
    }

    public override void Initialize(int nID, int nMasterDataID, params object[] param)
    {
        m_nID = nID;
        m_nMasterDataID = nMasterDataID;

        MasterData.Projectile masterProjectile = null;
        MasterDataManager.Instance.GetData<MasterData.Projectile>(m_nMasterDataID, ref masterProjectile);

        foreach(int nBehaviorID in masterProjectile.m_listBehaviorID)
        {
            IBehavior behavior = Factory.Instance.CreateBehavior(nBehaviorID);

            if (behavior == null)
            {
                Debug.LogWarning("behavior is null! nBehaviorID : " + nBehaviorID);
                continue;
            }

            behavior.Initialize(this, nBehaviorID, BaeGameRoom2.Instance.GetTickInterval());

            m_listBehavior.Add(behavior);
        }

        GameObject goEntityUI = ObjectPool.Instance.GetGameObject("CharacterModel/EntityUI");
        EntityUI entityUI = goEntityUI.GetComponent<EntityUI>();
        entityUI.Initialize(FBS.Data.EntityType.Projectile, nID, nMasterDataID);

        m_EntityUI = entityUI;
    }

    public override FBS.Data.EntityType GetEntityType()
    {
        return FBS.Data.EntityType.Projectile;
    }

    public override void NotifyGameEvent(IGameEvent gameEvent)
    {
    }

    public override void LateUpdateWorld(int nUpdateTick)
    {
        if (m_nDefaultBehaviorID != -1 && !IsBehavioring() && GetBehavior(m_nDefaultBehaviorID) != null)
        {
            GetBehavior(m_nDefaultBehaviorID).StartTick(nUpdateTick);
        }

        Move(m_vec3Velocity * BaeGameRoom2.Instance.GetTickInterval());
    }
}
